plugins/rbac-backend/src/role-manager/ancestor-search-memo.ts

```diff
@@ -0,0 +1,119 @@
+import { TokenManager } from '@backstage/backend-common';
+import { CatalogApi } from '@backstage/catalog-client';
+import { Entity } from '@backstage/catalog-model';
+
+import { alg, Graph } from '@dagrejs/graphlib';
+import { Logger } from 'winston';
+
+// AncestorSearchMemo - should be used to build group hierarchy graph for User entity reference.
+// It supports search group entity reference link in the graph.
+// Also AncestorSearchMemo supports detection cycle dependencies between groups in the graph.
+//
+export class AncestorSearchMemo {
+  private graph: Graph;
+
+  private tokenManager: TokenManager;
+  private catalogApi: CatalogApi;
+
+  private userEntityRef: string;
+
+  private allGroups: Entity[];
+
+  constructor(
+    userEntityRef: string,
+    tokenManager: TokenManager,
+    catalogApi: CatalogApi,
+  ) {
+    this.graph = new Graph({ directed: true });
+    this.userEntityRef = userEntityRef.split('/')[1];
+    this.tokenManager = tokenManager;
+    this.catalogApi = catalogApi;
+    this.allGroups = [];
+  }
+
+  isAcyclic(): boolean {
+    return alg.isAcyclic(this.graph);
+  }
+
+  findCycles(): string[][] {
+    return alg.findCycles(this.graph);
+  }
+
+  setEdge(parentEntityRef: string, childEntityRef: string) {
+    this.graph.setEdge(parentEntityRef, childEntityRef);
+  }
+
+  setNode(entityRef: string): void {
+    this.graph.setNode(entityRef);
+  }
+
+  hasEntityRef(groupRef: string): boolean {
+    return this.graph.hasNode(groupRef);
+  }
+
+  debugNodesAndEdges(log: Logger, userEntity: string): void {
+    log.debug(
+      `SubGraph edges: ${JSON.stringify(this.graph.edges())} for ${userEntity}`,
+    );
+    log.debug(
+      `SubGraph nodes: ${JSON.stringify(this.graph.nodes())} for ${userEntity}`,
+    );
+  }
+
+  getNodes(): string[] {
+    return this.graph.nodes();
+  }
+
+  async getAllGroups(): Promise<void> {
+    const { token } = await this.tokenManager.getToken();
+    const { items } = await this.catalogApi.getEntities(
```

Comment:
We need to check if catalog api really return all list groups. Theoretically it could support pagination and return list group limited by page size.

Reply:
Didn't think about the potential for pagination, will test to see if that is an issue and update accordingly.

Reply:
With about 520 groups, I don't seem to be running into a pagination issue. It looks like there is a separate property for the `GetEntitiesRequest` that would be used to [limit](https://backstage.io/docs/reference/catalog-client.getentitiesrequest/) the number of entities that are returned.

If you want me to test with more, I can. Just let me know if 520 groups is sufficient.